inference-engine/src/mkldnn_plugin/nodes/mkldnn_roll_node.cpp

```diff
@@ -0,0 +1,214 @@
+            rollImpl<PrecisionTrait<Precision::I16>::value_type>(input, shifts, axes, output);
+            break;
+        }
+        case sizeof(PrecisionTrait<Precision::I32>::value_type): {
+            rollImpl<PrecisionTrait<Precision::I32>::value_type>(input, shifts, axes, output);
+            break;
+        }
+        case sizeof(PrecisionTrait<Precision::I64>::value_type): {
+            rollImpl<PrecisionTrait<Precision::I64>::value_type>(input, shifts, axes, output);
+            break;
+        }
+        default:
+            IE_THROW() << "Roll has unsupported 'data' input precision: " << dataPrecision.name();
+    }
+}
+
+size_t MKLDNNRollNode::calculateShiftOffset(size_t dataOffset, size_t dimShift, size_t segmentSize, size_t dimSize) {
+    size_t pos = dataOffset / segmentSize % dimSize;
+    size_t shift = (pos + dimShift) % dimSize - pos;
+    return dataOffset + shift * segmentSize;
+}
+
+template <typename DataType>
+void MKLDNNRollNode::rollImpl(const Blob::CPtr &inputBlob, const Blob::CPtr &shiftsBlob, const Blob::CPtr &axesBlob, const Blob::Ptr &outputBlob) {
+    const auto *axes = axesBlob->cbuffer().as<const int32_t*>() + axesBlob->getTensorDesc().getBlockingDesc().getOffsetPadding();
+    const auto *shifts = shiftsBlob->cbuffer().as<const int32_t *>() + shiftsBlob->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+    const auto *input =
+            inputBlob->cbuffer().as<const DataType *>() + inputBlob->getTensorDesc().getBlockingDesc().getOffsetPadding();
+    auto *output = outputBlob->buffer().as<DataType *>() + outputBlob->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+    std::vector<size_t> shiftsVector(numOfDims, 0);
+    for (size_t dim = 0; dim < axesBlob->size(); ++dim) {
+        int32_t currentAxis = axes[dim] < 0 ? axes[dim] + numOfDims : axes[dim];
+        int32_t shiftSum = shiftsVector[currentAxis] + shifts[dim];
+        int32_t dimSize = shape[currentAxis];
+        shiftsVector[currentAxis] = (shiftSum % dimSize + dimSize) % dimSize;
+    }
+
+    const size_t blockSize = shape.back();
+    const size_t totalElements = inputBlob->size();
+    const size_t leftBlockSize = blockSize - shiftsVector.back();
+    const size_t rightBlockSize = blockSize - leftBlockSize;
+    const size_t elementSize = sizeof(DataType);
+
+    size_t nIterations = totalElements / blockSize;
+    parallel_for(nIterations, [&](size_t iter) {
+        size_t start = iter * blockSize;
+        size_t leftBlockStartOffset = start;
+        size_t rightBlockStartOffset = start + leftBlockSize;
+
+        size_t segmentSize = 1;
+        for (int dim = numOfDims - 1; dim >= 0; --dim) {
+            leftBlockStartOffset = calculateShiftOffset(leftBlockStartOffset, shiftsVector[dim], segmentSize, shape[dim]);
+            rightBlockStartOffset = calculateShiftOffset(rightBlockStartOffset, shiftsVector[dim], segmentSize, shape[dim]);
+            segmentSize *= shape[dim];
```

Comment:
Why `segmentSize` is calculated each time? Isn't it better to precalculate it into separate vector? BTW getTensorDesc().getBlockingDesc(),getStrides() should return the same vector.